sub-packages/bionemo-geometric/src/bionemo/geometric/base_featurizer.py

```diff
@@ -0,0 +1,105 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: LicenseRef-Apache2
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from abc import ABC, abstractmethod, abstractproperty
+from typing import List
+
+from rdkit.Chem import Atom, Mol
+
+
+class BaseFeaturizer(ABC):
+    """Abstract base featurizer class for all atom and bond featurization classes."""
+
+    @abstractproperty
+    def n_dim(self):
+        """Number of dimensions of compute feature."""
+        pass
+
+    def compute_features(self, mol: Mol) -> Mol:
```

Comment:
I will delete the modified reference to the same `mol` object. 

RE properties: 
I did not want the `Featurizer` classes to store any molecule-specific state. I want to instantiate one featurizer for all molecules. Storing molecule-specific information in the class would disallow me from using this pattern.
I like this pattern because some `Featurizers` require some one-time initialization (like [reading from property file](https://github.com/NVIDIA/bionemo-framework/blob/sveccham/impl_atom_bond_feats/sub-packages/bionemo-geometric/src/bionemo/geometric/atom_featurizers.py#L224) or [initializing SMARTS query molecules](https://github.com/NVIDIA/bionemo-framework/blob/sveccham/impl_atom_bond_feats/sub-packages/bionemo-geometric/src/bionemo/geometric/atom_featurizers.py#L314)) and I would like to do this once and not once per molecule.
Anyway this is possible?

Reply:
OK I see where you're coming from. Do you anticipate the bond and atom featurizers to iterate over all atoms and bonds? Or at least do that by default? In which case you could do something like.

```python
class MyFeaturizer(BaseFeaturizer): # where base has this abstract method
    def get_atom_features(self, mol, atomIndices: Optional[Iterable] = None):
        atomiterator = atomIndices if atomIndices else range(mol.GetNumAtoms())
        
        # Do any molecular preprocessing
        # ....

        res = np.zeros(self.n_dim,  len(atomIterator))
        for outIndex, atomIndex in enumerate(atomIterator):
            res[outIndex, :] = get_features(mol.GetAtomWithIdx(atomIndex))
        return res

```